clippy_lints/src/ptr.rs

```diff
@@ -153,14 +174,68 @@ impl<'tcx> LateLintPass<'tcx> for Ptr {
 
     fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
         if let ExprKind::Binary(ref op, l, r) = expr.kind {
-            if (op.node == BinOpKind::Eq || op.node == BinOpKind::Ne) && (is_null_path(l) || is_null_path(r)) {
+            if (op.node == BinOpKind::Eq || op.node == BinOpKind::Ne) && (is_null_path(cx, l) || is_null_path(cx, r)) {
                 span_lint(
                     cx,
                     CMP_NULL,
                     expr.span,
                     "comparing with null is better expressed by the `.is_null()` method",
                 );
             }
+        } else {
+            check_invalid_ptr_usage(cx, expr);
+        }
+    }
+}
+
+fn check_invalid_ptr_usage<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+    // (fn_path, arg_idx, num_args) -
+    //   `arg_idx` is the `arg` position where null would cause U.B.
+    //   `num_args` is the numbers of arguments of the function
+    const INVALID_NULL_PTR_USAGE_TABLE: [(&[&str], usize, usize); 20] = [
+        (&paths::SLICE_FROM_RAW_PARTS, 0, 2),
+        (&paths::SLICE_FROM_RAW_PARTS_MUT, 0, 2),
+        (&paths::PTR_COPY, 0, 3),
+        (&paths::PTR_COPY, 1, 3),
```

Comment:
You can make `arg_idx` a list so that there are no duplicate paths. `num_args` is not really necessary.

```suggestion
        (&paths::PTR_COPY, [0, 1]),
```